Mathlib/GroupTheory/GroupAction/Blocks.lean

```diff
@@ -76,275 +75,229 @@ end orbits
   hB.preimage C.inclusion
 
-theorem IsBlock.iff_subtype_val {C : SubMulAction G X} {B : Set C} :
-    IsBlock G B ↔ IsBlock G (Subtype.val '' B : Set X) := by
-  simp only [IsBlock.def_one]
-  apply forall_congr'
-  intro g
-  rw [← SubMulAction.inclusion.coe_eq, ← image_smul_set _ _ _ C.inclusion g B,
-    ← Set.image_eq_image Subtype.coe_injective]
-  apply or_congr Iff.rfl
-  simp only [Set.disjoint_iff, Set.subset_empty_iff, Set.image_eq_empty,
-    ← C.inclusion_injective.injOn.image_inter (Set.subset_univ _) (Set.subset_univ _)]
-
-theorem IsBlock.iff_top (B : Set X) :
-    IsBlock G B ↔ IsBlock (⊤ : Subgroup G) B := by
-  simp only [IsBlock.def_one]
-  constructor
-  · intro h g; exact h g
-  · intro h g; exact h ⟨g, Subgroup.mem_top g⟩
+theorem isBlock_subtypeVal {C : SubMulAction G X} {B : Set C} :
+    IsBlock G (Subtype.val '' B : Set X) ↔ IsBlock G B := by
+  refine forall₂_congr fun g₁ g₂ ↦ ?_
+  rw [← SubMulAction.inclusion.coe_eq, ← image_smul_set, ← image_smul_set, ne_eq,
+    Set.image_eq_image C.inclusion_injective, disjoint_image_iff C.inclusion_injective]
 
-/-- The intersection of two blocks is a block -/
-theorem IsBlock.inter {B₁ B₂ : Set X} (h₁ : IsBlock G B₁) (h₂ : IsBlock G B₂) :
-    IsBlock G (B₁ ∩ B₂) := by
-  rw [IsBlock.def_one]
-  intro g
-  rw [Set.smul_set_inter]
-  cases' h₁.smul_eq_or_disjoint g with h₁ h₁
-  · cases' h₂.smul_eq_or_disjoint g with h₂ h₂
-    · left; rw [h₁, h₂]
-    right
-    apply Disjoint.inter_left'; apply Disjoint.inter_right'
-    exact h₂
-  · right
-    apply Disjoint.inter_left; apply Disjoint.inter_right
-    exact h₁
-
-/-- An intersection of blocks is a block -/
-theorem IsBlock.iInter {ι : Type*} {B : ι → Set X} (hB : ∀ i : ι, IsBlock G (B i)) :
-    IsBlock G (⋂ i, B i) := by
-  by_cases hι : (IsEmpty ι)
-  · -- ι = ∅, block = univ
-    suffices (⋂ i : ι, B i) = Set.univ by simpa only [this] using isBlock_univ X
-    simpa only [Set.iInter_eq_univ] using (hι.elim' ·)
-  rw [IsBlock.def_one]
-  intro g
-  rw [Set.smul_set_iInter]
-  by_cases h : ∃ i : ι, Disjoint (g • B i) (B i)
-  · right
-    obtain ⟨j, hj⟩ := h
-    refine Disjoint.mono ?_ ?_ hj <;> apply Set.iInter_subset
-  · left
-    simp only [not_exists] at h
-    have : ∀ i : ι, g • B i = B i := fun i => ((hB i).smul_eq_or_disjoint g).resolve_right (h i)
-    rw [Set.iInter_congr this]
-
-theorem IsBlock.of_subgroup_of_conjugate {B : Set X} {H : Subgroup G} (hB : IsBlock H B) (g : G) :
+theorem IsBlock.of_subgroup_of_conjugate {H : Subgroup G} (hB : IsBlock H B) (g : G) :
     IsBlock (Subgroup.map (MulEquiv.toMonoidHom (MulAut.conj g)) H) (g • B) := by
-  rw [IsBlock.def_one]
+  rw [isBlock_iff_smul_eq_or_disjoint]
   intro h'
   obtain ⟨h, hH, hh⟩ := Subgroup.mem_map.mp (SetLike.coe_mem h')
   simp only [MulEquiv.coe_toMonoidHom, MulAut.conj_apply] at hh
   suffices h' • g • B = g • h • B by
     simp only [this]
-    apply (hB.smul_eq_or_disjoint ⟨h, hH⟩).imp
+    apply (hB.smul_eq_or_disjoint (g := ⟨h, hH⟩)).imp
```

Comment:
This makes it seem like making `g` implicit was a bad idea - there are no cases where you saved an underscore (or anything else, for that matter), but there is a case where you needed to specify it explicitly. What's the justification for this?

Reply:
Great!